syntax = "proto3";

package idbg;

message Node {
    int32 id = 1;

    repeated string labels = 2;

    bool is_func = 3;

    repeated double data = 4; // maximum precision for laziness

    string repr = 5;
}

message Edge {
    int32 parent = 1;

    int32 child = 2;

    int32 order = 3;
}

message Graph {
    repeated Node nodes = 1;

    repeated Edge edges = 2;
}

message GraphUpdateRequest {
    Graph payload = 1;
}

message GraphUpdateResponse {
    enum Status {
        OK = 0;
        BAD_INPUT = 1; // basically 400's error
        INTERNAL_ERR = 2; // basically 500's error
        UNKNOWN = 3; // something bad
    }

    Status status = 1;

    string message = 2;
}

service InteractiveGrapher {
    rpc UpdateGraph (GraphUpdateRequest) returns (GraphUpdateResponse);
}
